BeneficiariesFile now expects a rational number, which can be scaled …

jy14898 opened the pull request:
…by the --decimal-places flag

Closes https://github.com/mlabs-haskell/CardStarter-bulk-airdrop-script/issues/18.

Should we warn on truncation?

jy14898 committed:
BeneficiariesFile now expects a rational number, which can be scaled by the --decimal-places flag

jy14898 committed:
Apply format

samuelWilliams99 commented:
Truncation is an annoying one - i think we need to be REALLY explicit here.
This will extend the task a little bit, but I think the safest bet is to completely fail parsing on truncation - UNLESS, we have a flag that allows it.
This flag should be very explicit that is always rounds down.
So, to summarise:
- Add a truncate flag - make it obvious that it rounds down (using the help text)
- If we try to truncate a value and this flag is not set, fail parsing.

samuelWilliams99 commented on src/BeneficiariesFile.hs:
```diff
@@ -50,16 +51,19 @@ parseBeneficiary conf = toBeneficiary . words
-    makeBeneficiary :: Text -> Either Text Integer -> Either Text AssetClass -> Either Text Beneficiary
-    makeBeneficiary addr eAmt eAc = Beneficiary <$> parseAddress conf.usePubKeys addr <*> eAmt <*> eAc
+    scaleAmount :: Ratio Integer -> Integer
+    scaleAmount r = numerator r * conf.decimalPlaces `div` denominator r
```
does multiplying the numerator by the decimal places do what we want? should it not be something like `10^decimal places`?

jy14898 committed:
Fix scaleAmount

jy14898 committed:
Moved from Rational to Scientific for drop amounts. Added a truncation flag and test for truncation.

jy14898 committed:
Update README.md

jy14898 commented:
I've moved from Rational to Scientific for representing pre-integer token values because:
 - Scientific more naturally represents what we're doing here (c * (10^e))
 - Actually parses correctly

I've added the `--truncate` flag to allow dropping change when rounding down. I've also added a test which checks if the truncation occurs only when it should (regardless of the flag setting).

jy14898 committed:
Remove full stop from truncate long help text

jy14898 committed:
Move scientificToInteger from Either to Maybe

jy14898 committed:
Fix building

jy14898 committed:
Simplify scientificToInteger

jy14898 commented on src/BeneficiariesFile.hs:
```diff
@@ -50,16 +51,32 @@ parseBeneficiary conf = toBeneficiary . words
+    b10e = base10Exponent s
+
+    -- For truncated to be evaluated, the exponent must have been negative
+    truncated = Just . (*) (10 ^^ b10e) . fromInteger <$> coefficient s `divMod` (10 ^ negate b10e)
```
I've just pushed another go at it, no longer using point free/tuple functor instance

samuelWilliams99 replied:
Thats looks better, thanks! Happy to merge when this passes CI
